### PR TITLE
KAFKA-13115 Update doSend doc about possible blocking

### PR DESCRIPTION
**Pull request by IvanVas**
### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by IvanVas** [comment]
@jsancio Hi! Could you help?
How should I proceed with the PR?

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java [hunk 1]
```diff
@@ -778,9 +778,10 @@ public Future<RecordMetadata> send(ProducerRecord<K, V> record) {
     /**
      * Asynchronously send a record to a topic and invoke the provided callback when the send has been acknowledged.
      * <p>
-     * The send is asynchronous and this method will return immediately once the record has been stored in the buffer of
+     * The send is asynchronous and this method will return immediately (except for getting the topic metadata) once the record has been stored in the buffer of
```
Comment (jsancio) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: @IvanVas, it looks like the code does block waiting for metadata information on the cluster but only if it doesn't have the necessary information on the topic. In other works the producer only blocks on the metadata response, if this is the first record being sent to the cluster by this client for the given topic.

How about encapsulating that in the comment?

**Review by jsancio — COMMENTED** [review]
Thanks for the documentation update and apologies for the delayed response.

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java (hunk 1 above)
Comment (IvanVas) [reply clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: > @IvanVas, it looks like the code does block waiting for metadata information on the cluster but only if it doesn't have the necessary information on the topic. In other works the producer only blocks on the metadata response, if this is the first record being sent to the cluster by this client for the given topic.
> 
> How about encapsulating that in the comment?

Thanks for the clarification! What do you think about the updated description?

File: clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java [hunk 2]
```diff
@@ -778,9 +778,13 @@ public Future<RecordMetadata> send(ProducerRecord<K, V> record) {
     /**
      * Asynchronously send a record to a topic and invoke the provided callback when the send has been acknowledged.
      * <p>
-     * The send is asynchronous and this method will return immediately once the record has been stored in the buffer of
+     * The send is asynchronous and this method will return immediately (except for getting the topic metadata if it 
+     * doesn't have the necessary information about the topic) once the record has been stored in the buffer of
      * records waiting to be sent. This allows sending many records in parallel without blocking to wait for the
      * response after each one.
+     * Note, that getting metadata part of a method is blocking. This happens for the first record being sent to 
+     * the cluster by this client for the given topic. In this case it will block (for up to maxBlockTimeMs) if
```
Comment (jsancio) [inline clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java]: Let's use `max.block.ms`. `maxBlockTimeMs` is an internal field that the user doesn't know about.

**Review by jsancio — COMMENTED** [review]
@IvanVas, I was looking at the implementation some more and it looks like send can also block trying to allocate a buffer. For example,

```java
            // Append the record to the accumulator.  Note, that the actual partition may be                                                                                                                                                            
              // calculated there and can be accessed via appendCallbacks.topicPartition.                                                                                                                                                                 
              RecordAccumulator.RecordAppendResult result = accumulator.append(record.topic(), partition, timestamp, serializedKey,                                                                                                                       
                      serializedValue, headers, appendCallbacks, remainingWaitMs, abortOnNewBatch, nowMs, cluster); 
```
in `doSend`.

This is the documentation from RecordAccumulator:
```java
@param maxTimeToBlock The maximum time in milliseconds to block for buffer memory to be available
...
public RecordAppendResult append(
...
```

Maybe in the first paragraph "The send is asynchronous..." we can mention that in some infrequent cases this method can block up to `max.block.ms`.

In the second paragraph we can detail when this will happen; 1) waiting for metadata for unknown topic, 2) allocating a buffer if  buffer pool doesn't have any free buffers. For the buffer pool case we can think about mentioning how the user can controller the buffer pool size (`buffer.memory` and `batch.size`). Note that this is also mentioned in the class-level documentation for `KafkaProducer`

What do you think?

**Review by jsancio — APPROVED** [review]
LGTM. Thanks for the contribution.

**Pull request merged**